pkg/controllers/updaterun/execution_test.go

```diff
@@ -398,3 +407,522 @@ func TestBuildApprovalRequestObject(t *testing.T) {
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "test-update-run",
+					Generation: 1,
+				},
+				Spec: placementv1beta1.UpdateRunSpec{
+					PlacementName:         "test-placement",
+					ResourceSnapshotIndex: "1",
+				},
+				Status: placementv1beta1.UpdateRunStatus{
+					StagesStatus: []placementv1beta1.StageUpdatingStatus{
+						{
+							StageName: "test-stage",
+							Clusters: []placementv1beta1.ClusterUpdatingStatus{
+								{
+									ClusterName: "cluster-1",
+									// No conditions - cluster has not started updating yet.
+								},
+							},
+						},
+					},
+					UpdateStrategySnapshot: &placementv1beta1.UpdateStrategySpec{
+						Stages: []placementv1beta1.StageConfig{
+							{
+								Name:           "test-stage",
+								MaxConcurrency: &intstr.IntOrString{Type: intstr.Int, IntVal: 1},
+							},
+						},
+					},
+				},
+			},
+			bindings: []placementv1beta1.BindingObj{
+				&placementv1beta1.ClusterResourceBinding{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "binding-1",
+						Generation: 1,
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster:        "cluster-1",
+						ResourceSnapshotName: "test-placement-1-snapshot",        // Already synced.
+						State:                placementv1beta1.BindingStateBound, // Already Bound.
+					},
+					Status: placementv1beta1.ResourceBindingStatus{
+						Conditions: []metav1.Condition{
+							{
+								Type:               string(placementv1beta1.ResourceBindingRolloutStarted),
+								Status:             metav1.ConditionTrue,
+								ObservedGeneration: 1,
+								Reason:             condition.RolloutStartedReason,
+							},
+							{
+								Type:               string(placementv1beta1.ResourceBindingApplied),
+								Status:             metav1.ConditionFalse,
+								ObservedGeneration: 1,
+								Reason:             condition.ApplyFailedReason,
+							},
+						},
+					},
+				},
+			},
+			interceptorFunc: nil,
+			wantErr:         errors.New("cluster updating encountered an error at stage"),
+			wantWaitTime:    0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			scheme := runtime.NewScheme()
+			_ = placementv1beta1.AddToScheme(scheme)
+
+			var fakeClient client.Client
+			objs := make([]client.Object, len(tt.bindings))
+			for i := range tt.bindings {
+				objs[i] = tt.bindings[i]
+			}
+			if tt.interceptorFunc != nil {
+				fakeClient = interceptor.NewClient(
+					fake.NewClientBuilder().WithScheme(scheme).WithObjects(objs...).Build(),
+					*tt.interceptorFunc,
+				)
+			} else {
+				fakeClient = fake.NewClientBuilder().WithScheme(scheme).WithObjects(objs...).Build()
+			}
+
+			r := &Reconciler{
+				Client: fakeClient,
+			}
+
+			// Execute the stage.
+			waitTime, gotErr := r.executeUpdatingStage(ctx, tt.updateRun, 0, tt.bindings, 1)
+
+			// Verify error expectation.
+			if (tt.wantErr != nil) != (gotErr != nil) {
```

Comment:
Also validate if gotErr is `errAborted1?

Reply:
We are already checking the error messages for both about cases,

- https://github.com/kubefleet-dev/kubefleet/pull/323/files#diff-9ca2857cf67fef84221c2a0cb673a3845e8197afa9fcf08469ab44a6f33b5f1eR421
- https://github.com/kubefleet-dev/kubefleet/pull/323/files#diff-9ca2857cf67fef84221c2a0cb673a3845e8197afa9fcf08469ab44a6f33b5f1eR519